### ISSUE TITLE
[Request] Picture-in-picture for webcam stream (to allow it to be always-on-top)

### ISSUE THREAD

**Issue by DanTup** [OctoPrint/OctoPrint#2956]
I'm loving being able to keep an eye on my prints without running upstairs (especially after a mishap with a print coming loose earlier)! However it'd be even better if I could have the camera stream visible on my laptop always on top of whatever else I'm doing (so I can keep an eye on it while browsing the web, etc.).

A recent version of Chrome added support for this, called "Picture in picture". You can see it on YouTube, be right-clicking on a video twice (?) and then selecting "Picture-in-picture". You get an overlay that now persists (at the desktop level) so you can switch to do other things (and move it around/resize it/etc.):

https://www.theverge.com/2018/10/19/17999482/chrome-70-picture-video-youtube-android-windows-mac-linux

It'd be awesome of Octoprint could support this (note: I have no idea how technically complicated is, but I hope Chrome is doing most of the hard work).

**Comment by jneilliii** [comment]
That's a neat plugin concept.  It seems to be specific to video, and unfortunately the most common streamer for OctoPrint video is an mjpg stream, which I don't believe lends itself to embedding in a video tag (required by the Chrome side). You could probably achieve your same end goal by using the webcam stream url in an "always on top" browser window if you could find an app to do that..

**Comment by DanTup** [comment]
> unfortunately the most common streamer for OctoPrint video is an mjpg stream, which I don't believe lends itself to embedding in a video tag

Ah, that's a shame.

> You could probably achieve your same end goal by using the webcam stream url in an "always on top" browser window if you could find an app to do that.

There are some Chrome extensions to do it, but they generally require god-permissions and are from unknown authors so probably shouldn't be trusted (the Google-published extension linked in the article above I think still only works for video tags, it just bypasses the need for the author to have enabled it specifically).

Slightly off-topic, but is there a way to change the cam to use something video-tag-friendly? (I did search, but can't find anything obvious).

**Comment by jneilliii** [comment]
> Slightly off-topic, but is there a way to change the cam to use something video-tag-friendly? (I did search, but can't find anything obvious).

There's been several posts about using different stream technologies, using H264, etc. that would in theory work with video tags, but as of yet I have not seen any published working solutions.  I have a couple of plugins (RTMPStreamer and YoutTubeLive) that would allow to stream to external sources, one of which is specifically YouTube, but it's probably not realtime enough to suite your needs.  There is a ~30-45 second delay for the re-encoding process.

**Comment by DanTup** [comment]
A delay wouldn't be terrible (finding out it's gone bad 45 seconds later is probably still sooner than if I have to tab back periodically to check), though a local stream would be better (not only is the latency better, it's not needlessly burning broadband bandwidth).

(FWIW, if the re-encoding delay you're talking about is on the YT server side, [Mixer](https://mixer.com/) had sub-second latency when I compared them, and I suspect streaming to them is in large similar to streaming to YT)

**Comment by reloxx13** [comment]
another idea: connect multiple octoprint instances (entering local ips in settings) and than show a new tab with all webcams, doubleclick to fullscreen one.

**Comment by tedder** [comment]
Yeah, I was looking at it. The best option seems to be using WebRTC, but it's hard to find JS examples that are for a one-way stream. But it works.

I have five jpegstream-type vids embedded on one page. It kills my machine's cpus when I leave that tab open, it's awful.

**Comment by mvaneijgen** [comment]
Had the same question but from a macOS point of view https://discourse.octoprint.org/t/picture-in-picture-mode-for-the-webcam-view-on-macos-ios/5258 

Probably think it requires the same technology (eg the use of a `<video>` tag).

**Comment by nixmeer** [comment]
I'd love to see this feature for macOS!!

**Comment by TheShanMan** [comment]
I just got OctoPrint set up for the first time and one of the first things I wanted to do was picture in picture with the video stream. I'd love to see this. :)

**Comment by jneilliii** [comment]
You guys might be able to achieve this with the experimental HLS support provided in the latest octopi 0.18 image. It utilizes a video container [here](https://github.com/OctoPrint/OctoPrint/blob/361162aac2ed906b393d05bc4a69d0e41a8a4457/src/octoprint/templates/tabs/control.jinja2#L4).

**Comment by TheShanMan** [comment]
I have 0.18 installed so unless there's some configuring required, that hasn't enabled it. If I right click the video it gives me an image menu (such as "open image in new tab").

**Comment by jneilliii** [comment]
There is. In octopi.txt you have to uncomment the option at the bottom and set it to hls.

**Comment by cp2004** [comment]
It doesn't have picture in picture yet, the suggestion is more that someone could work on it - since picture in picture is only supported by the video tag, which is only used with HLS

**Comment by jneilliii** [comment]
if using octopi 0.18 image you can just change the setting at the bottom of octopi.txt to hls.

```
### EXPERIMENTAL
# Support for different streamer types.
#
# Available options:
#   mjpeg [default] - stable MJPG-streamer
#   hls - experimental FFMPEG HLS streamer
camera_streamer=hls
```

**Comment by cp2004** [comment]
With HLS & WebRTC I would call this 'addressed' in 1.8.0. The default webcam streaming on OctoPi will remain as mjpg for now as it is the most accessible, while work is done to find a suitable, reliable WebRTC streamer that can be implemented in OctoPi images.
